### ISSUE TITLE
disableClickPropagation fails after double tapping map on iOS

### ISSUE THREAD

**Issue by ted-piotrowski** [Leaflet/Leaflet#7422]
![IMG_2DE51D162B81-1](https://user-images.githubusercontent.com/2023571/105534292-216a3d00-5ca2-11eb-8a35-61e7494a408d.jpeg)

**Steps to reproduce**
- Open [https://6v1ht.csb.app](https://6v1ht.csb.app) on iOS device
- **Double tap** on upper 50% of map to zoom
- **Single tap** on overlay element  in lower 50% of map

**Expected behavior**
- **Single tap** on overlay element should not do anything because it is a `<div>` with `disableClickPropagation`

**Current behavior**
- Map zooms in

**Environment**
- Leaflet version: 1.7.1
- Browser (with version): iOS Safari
- OS/Platform (with version): iOS 14.3

**Additional context**
- Also broken on [https://shademap.app](https://shademap.app) when you attempt to use slider at bottom of map.
- Thank you for maintaining this amazing library. I am investigating the bug and hoping to narrow down the cause. Any guidance is greatly appreciated.

**Minimal example reproducing the issue**
https://codesandbox.io/s/epic-stonebraker-6v1ht?file=/src/index.js

**Comment by johnd0e** [comment]
Try to set map option `tap: false`.
(And show exact code where you do it)

**Comment by ted-piotrowski** [comment]
Setting `tap: false` has no effect on the behavior.

```
var map = L.map("map", { tap: false }).setView([51.505, -0.09], 13);

L.tileLayer("https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png", {
  attribution:
    '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
}).addTo(map);

const el = L.DomUtil.get("overlay");
L.DomEvent.disableClickPropagation(el);
```

**Comment by ted-piotrowski** [comment]
Found the cause.

DoubleTap handler listens for `touchstart` events and uses a timer to determine if two `touchstart` events occur within a specified `delta`. If they happen within a `delta`, a `doubleTap` flag is set to true. On a `touchend` event, if the `doubleTap` flag is set to true a `dblclick` event is dispatched and the timer is reset, HOWEVER, the `doubleTap` flag remains true until another `touchstart` event occurs. 

The assumption here is that there CANNOT be a `touchend` event that's not preceded with be a `touchstart`.

Unfortunately, if an element has `stopClickPropagation` enabled, the element no longer propagates `mousedown touchstart dblclick` but will still propagate `touchend` to the map.

The other way to fix this is to modify `stopClickPropagation` to prevent all `mousedown touchstart touchend dblclick` events from propagating but I think this approach is safer.

**Comment by johnd0e** [comment]
We have another ultimate solution: #7027. Please test.

**Comment by ted-piotrowski** [comment]
@johnd0e I can confirm #7027 solves this issue, but it creates a new problem: `disableClickPropagation` no longer stops propagation of double taps. 

To reproduce, overlay an element over a map and `disableClickPropagation` then double tap the element and notice that the map zooms in.

**Comment by johnd0e** [comment]
@ted-piotrowski 
Thank you for pointing this out!

Could you test new revision of PR?

**Comment by ted-piotrowski** [comment]
@johnd0e The new PR prevents propagation of double taps but creates a new problem:

Tap on a control with `disableClickPropagation` enabled (like the zoom (-/+)) then tap on the map. The map will zoom in. `simDblClick` is being called twice.

If you just tap on the map after page load `simDblClick` is only called once. It's only after tapping a `disableClickPropagation` element and then taping on the map the `simDblClick` is called twice.

**Comment by johnd0e** [comment]
> `simDblClick` is being called twice.

Unable to reproduce.

But there is still one edge case: if you make two taps promptly: 1) on control 2) on map.
This test case failed:
```js
	it.only('respects disableClickPropagation (case 2)', function () {
		var spyMap = sinon.spy();
		var map = L.map(el).setView([51.505, -0.09], 13);
		map.on('dblclick', spyMap);

		var spyCtrl = sinon.spy();
		var ctrl = L.DomUtil.create('div');
		L.DomEvent.disableClickPropagation(ctrl);
		var MyControl = L.Control.extend({
			onAdd: function () {
				return ctrl;
			}
		});
		map.addControl(new MyControl());
		L.DomEvent.on(ctrl, 'dblclick', spyCtrl);

		happen.click(ctrl, {detail: 1});
		clock.tick(100);
		happen.click(map._container, {detail: 1});

		expect(spyCtrl.notCalled).to.be.ok();
		expect(spyMap.notCalled).to.be.ok();
	});
```
Did you mean that case?
But there is only single "dblclick" anyway.

**Comment by ted-piotrowski** [comment]
@johnd0e I am also unable to reproduce using pure Leaflet library. 

I'm only seeing this behavior when using Leaflet React. After clicking on a disableClickPropagation control and then clicking on the map two `click` events are fired. One that isTrusted and one that is not. This might be caused by React synthetic events. I do not know enough to suggest to ignore untrusted clicks in `simDblClick`. I'm looking into it now and I will share any additional info I find. 

Since it's not your task to support React Leaflet, feel free to close this issue if you merge your fix.

**Comment by johnd0e** [comment]
Make sure that map `tap` handler is disabled.

**Comment by johnd0e** [comment]
> One that isTrusted and one that is not

It's easy fix by rejecting not trusted events, but you better to confirm event's source

**Comment by ted-piotrowski** [comment]
> Make sure that map `tap` handler is disabled.

Setting `tap: false` fixes the React Leaflet issue. Seems like Map.Tap.js was firing the simulated `click` event. Unsure why it only starts firing this after clicking on a `disableClickPropagation` element first.

**Comment by johnd0e** [comment]
See #7026

**Comment by ted-piotrowski** [comment]
Also a correction to my previous comment. This issue is also present in vanilla Leaflet with `tap: false` built from your branch. You can see the example here.

https://k9cc8.csb.app/ (Must use Chrome Dev tools with mobile emulation)

I assume that when #7026 is merged this won't be an issue anymore.

**Comment by johnd0e** [comment]
@ted-piotrowski 
Please confirm if you observe same behavior here: https://gist.githack.com/johnd0e/12cb0d6c2aead30ea7f2e17c5f9991be/raw/test-doubletap.html

If it is exactly the same, then #7026 cannot help.
Even more: native `dblclick` also induces map zoomin.

**Comment by ted-piotrowski** [comment]
Assuming the yellow square control or attribution has `disableClickPropagation` enabled then I cannot reproduce the above behavior. 

The only thing I see is that tapping the yellow square control or attribution four times rapidly triggers a simulated `dblclick`
